### PR TITLE
Blocking Refactor I

### LINKED ISSUE #1300: bug: attack animations runs after blocking
**Describe the bug**
When the sprite has custom attack sprites (attack, weapon or shoot), this animation is played after blocking.

**Version Encountered**
0.7.1.22

**To Reproduce**
1. Make extra sprite animation
2. Equip a shield and release
3. After release de block button see error

**Expected behavior**
Do not run attack animation, if not attacking

**Screenshots**
https://www.ascensiongamedev.com/resources/filehost/04fa0e69f0facca18722f5342a14343b.mp4

**Logs**
If there are error logs please attach them below.

**Operating System:**
Windows

**Additional context**

### PR DESCRIPTION
**Pull request by WeylonSantana**
Resolves #1300

I believe that the block was created to act as a "status", the player while holding the button is in a "defense state" until released, I changed that.

I'm using the attack logic as a base to modify the blocking, now, it is working similar to the attack, as an action and not as a "state", so while holding the block button, it keeps making several attempts to block, as it works in attack.
Basically, the player either attacks or defends, never both, and [this fixes the attack animation playing after blocking.](https://github.com/AscensionGameDev/Intersect-Engine/issues/1300)

I just changed the logic a little bit of how it works to make it look like the attack with "speed modifier".

### CODE REVIEW

File: Intersect.Client/Core/Input.cs [hunk 1]
```diff
@@ -185,7 +185,10 @@ public static void OnKeyPressed(Keys modifier, Keys key)
                                         break;
 
                                     case Control.Block:
-                                        Globals.Me?.TryBlock();
+                                        if(Globals.Me.IsBlocking == false)
```
Comment (lodicolo) [inline Intersect.Client/Core/Input.cs:188]: Instead of checking outside of TryBlock, this check belongs inside of TryBlock:

```
if (IsBlocking)
{
    return true;
}
```

File: Intersect.Client/Entities/Player.cs [hunk 2]
```diff
@@ -242,6 +242,15 @@ public override bool Update()
...
                     }
                 }
+
+                //Holding block button for "auto blocking"
+                if(Controls.KeyDown(Control.Block))
+                {
+                    if(!IsBlocking)
+                    {
+                        TryBlock();
+                    }
```
Comment (lodicolo) [inline Intersect.Client/Entities/Player.cs:252]: See above `IsBlocking` check comment

File: Intersect.Client/Networking/PacketHandler.cs [hunk 3]
```diff
@@ -1014,10 +1014,30 @@ public void HandlePacket(IPacketSender packetSender, EntityAttackPacket packet)
...
+                    if (attackTimer > -1)
+                    {
+                        en.AttackTimer = Timing.Global.Ticks / TimeSpan.TicksPerMillisecond + attackTimer;
+                        en.IsBlocking = true;
+                    }
+                    else
+                    {
+                        en.IsBlocking = false;
+                    }
+                }
```
Comment (lodicolo) [inline Intersect.Client/Networking/PacketHandler.cs:1040]: I think this can be greatly simplified to something like:

```csharp
var isSelf = en == Globals.Me;
if (isSelf)
{
    en.IsBlocking = packet.IsBlocking;
}

if (attackTimer > -1)
{
    en.AttackTimer = Timing.Global.Ticks / TimeSpan.TicksPerMillisecond + attackTimer;
    if (!isSelf)
    {
        en.AttackTime = attackTimer;
    }
}
```

File: Intersect.Server/Entities/Entity.cs [hunk 4]
```diff
@@ -346,6 +346,13 @@ public virtual void Update(long timeMs)
                     {
                         status.TryRemoveStatus();
                     }
+
+                    //Blocking timers
+                    if(AttackTimer < Timing.Global.Milliseconds && Blocking)
```
Comment (lodicolo) [inline Intersect.Server/Entities/Entity.cs:351]: Make this be `Blocking && ...`.

File: Intersect.Server/Entities/Entity.cs [hunk 5]
```diff
@@ -812,7 +819,7 @@ public virtual float GetMovementTime()
             var time = 1000f / (float) (1 + Math.Log(Stat[(int) Stats.Speed].Value()));
             if (Blocking)
             {
-                time += time * Options.BlockingSlow;
+                time += time * (float)Options.BlockingSlow;
```
Comment (lodicolo) [inline Intersect.Server/Entities/Entity.cs:822]: What effect does this change have?

File: Intersect.Server/Networking/PacketSender.cs [hunk 6]
```diff
@@ -955,9 +955,9 @@ public static void SendEntityDir(Entity en)
         }
 
         //EntityAttackPacket
-        public static void SendEntityAttack(Entity en, int attackTime)
+        public static void SendEntityAttack(Entity en, int attackTime, bool isBlocking = false)
```
Comment (lodicolo) [inline Intersect.Server/Networking/PacketSender.cs:958]: I think this parameter should always be explicit, no default `= false` here.

**Review by lodicolo — CHANGES_REQUESTED** [review]

File: Intersect.Server/Entities/Entity.cs (hunk 5 above)
Comment (WeylonSantana) [reply Intersect.Server/Entities/Entity.cs:822]: I think I saw that on the client side it was like this, so I left both the same, client and server, nothing really changes

File: Intersect.Server/Networking/PacketSender.cs (hunk 6 above)
Comment (WeylonSantana) [reply Intersect.Server/Networking/PacketSender.cs:958]: The intention is to make blocking separate from the attack, later on (maybe after this merge), so it's implicit only for the blocking use case, to make it easier to remove when separating from the attack

File: Intersect.Client/Networking/PacketHandler.cs [hunk 7]
```diff
@@ -1014,29 +1014,18 @@ public void HandlePacket(IPacketSender packetSender, EntityAttackPacket packet)
...
+            var isSelf = en == Globals.Me;
+            if (isSelf)
             {
-                //attack handler
-                if (attackTimer > -1 && en != Globals.Me)
-                {
-                    en.AttackTimer = Timing.Global.Ticks / TimeSpan.TicksPerMillisecond + attackTimer;
-                    en.AttackTime = attackTimer;
-                }
+                en.IsBlocking = packet.IsBlocking;
```
Comment (lodicolo) [inline Intersect.Client/Networking/PacketHandler.cs:1020]: I just realized that my suggestion is not actually compatible with your logic, this is my bad.

I think this actually needs to be (in order to match the functionality you were adding):

```
if (isSelf && packet.IsBlocking)
{
    en.IsBlocking = attackTimer > -1;
}
```

**Review by lodicolo — CHANGES_REQUESTED** [review]

**Review by lodicolo — APPROVED** [review]

**Review by Arufonsu — APPROVED** [review]
